### ISSUE TITLE
libimaglink: Annotations for internal links?

### ISSUE THREAD

**Issue by matthiasbeyer** [matthiasbeyer/imag#765]
Would it be a good idea to be able to create annotations for internal links?

Use case: You have three mail objects in the database: 
- `Mail 1` and `Mail 2` should be linked because they are in one mail thread
- `Mail 1` and `Mail 3` should be linked because they are from the same author

---
## Benefits
- Currently we have the limitation that we can only link. With annotated links, we could create "linked-author" and "link-thread" links in the `libimagmail` and speed up the search for "Give me all mails from the thread where this mail is in"-like searches _a lot_.
## Drawbacks
- This makes links complex. The interface for `libimaglink` would get additional features and would be more complex, I don't see any more drawbacks by now.
## Implementation (ideas)

I have two ideas for how to implement this
### Alter the `link` header entry.

The first idea would simply add fields to the `[link]` section in the header. That's rather easy, not backwards compatible (we could make it backwards-compatible, though I see not that much value in it, as we are pre-1.\* releases), but should work.
### Create intermediate entries

This would be more complex. We would create intermediate link objects which contain annotations then.

```
A <==== <C with annotation> ====> B
```

Drawbacks would be that this would be really complex to implement.

---

I like the first approach much more, but we might have a discussion whether the second one would be doable or something.

@TheNeikos @neithernut @mario-kr What do you think.
1. Do we need annotations for links?
2. How to implement them?

If yes, I would make this a hard-required issue for milestone 0.3.0, I guess. This would then move the mail library `libimagmail` from 0.2.0 to 0.3.0 (or later) as this would be a crucial feature for the mail library, IMO.

**Comment by neithernut** [comment]
For the use case you mentioned, I don't think you need annotated links. The author, the thread and other data can be easily retrieved from the mail itself. Duplicating that information somewhere else could even produce new problems (since you have to keep them synced). It *could* speed up some searches, but I'd rather use a cache for speeding up things, for which an obvious way exists to keep things up-to-date, and which may be discarded at any time (e.g. if it's stale).

That being said, there might still be other use cases for which the concept might be of use. An Annotation might, for example, give a user an idea of why the link was created. Also, links carrying module specific information might still bring some advantages.

In the first case, the interface ought to be simple. A user would add a small comment to a link, which might be stored in an additional field in the `link` header. It sure would be the least complex way to implement this.
In the second case, a module will most certainly want to store some structured information with the link. Extending the `link` entry header to store arbitrary data would be rather awkward, imo. Modules should not fiddle with (standard) headers, not belonging to themselves. Also, the standard headers should follow a rather strict format (amongst other things, this enables us to keep the parser clean). Also, in such a use-case, we should not assume that, conceptually, such a link would be only of interest between **two** entries. Maybe a module "links" sets of entries together. Extending the link facility to support this would clearly be contrary to the original design.

Links should be kept simple. If users want to (manually) comment on links, an additional field could be added to the link header.
However, if a module wants to store structured data, the right approach, in my opinion, would be for the module to create a special entry for itself and manage the links to all the entries affected. That entry only has to be processed by the module, so the internal structure is quite free in its design and we don't have to deal with the structure *anywhere* else. The core linking code itself would not deal with that internal data but only be invoked for managing the links. So there is no need to touch it at all.

**Comment by matthiasbeyer** [comment]
> The author, the thread and other data can be easily retrieved from the mail itself.

Yes, but the mentioned usecase was to retrieve _related_ links, and if you have direct links in the store, you can do this much faster as loading possibly _thousands_ of mails from a Maildir, just to find out that there are no related mails (read: more mails in the same thread) (or only few).

> Duplicating that information somewhere else could even produce new problems (since you have to keep them synced).

This is surely a problem somehow. Maybe we should bench this ... Reading thousands of mails vs. reading thousands of imag entries.

In the long run, I'm sure the latter is better, as imag may evolve from a one-shot program to a long-running program (read: curses interface which gets started once and then gets used and that's it).

I could live with a read-thousands-of-mails-per-imag-call solution if we get the interface right and can switch to an imag-caches-this-stuff-implementation later on. I'm pretty sure that is not that much of a problem.

TL;DR: If reading meta-information from FS is fast enough (for thousands of mails), I could live with that.

---

> In the first case, the interface ought to be simple. A user would add a small comment to a link, which might be stored in an additional field in the link header. It sure would be the least complex way to implement this.

Sounds good. When we're at it: It might be useful to add some more metadata to a link, like creation date for example. We might find some more.

The implementation might change the header format in this specific way:

```toml
# Before:
links = [ "foo/bar", "baz/bar" ]

# After
links = [ 
  { link = "foo/bar" }, 
  { link = "baz/bar", 
    annotations = { 
      comment = "Some comment by the user",
      created = "2016-09-09 09:09:09",
    }
  }
]
```

(Not sure whether this is correct TOML, I guess not.

---

> However, if a module wants to store structured data, the right approach, in my opinion, would be for the module to create a special entry for itself and manage the links to all the entries affected. That entry only has to be processed by the module, so the internal structure is quite free in its design and we don't have to deal with the structure anywhere else. The core linking code itself would not deal with that internal data but only be invoked for managing the links. So there is no need to touch it at all.

I see the advantage and I like the approach, **and**:

1. We can re-use the `libimagentrylink` library and add some more core-code which does basic linking, the existing interface would then be the standard-implementation for linking with the `[link]` toml header (what is happening right now wouldn't change, except the new features mentioned above).
    1. Adavtage would be that the existing interface would not be changed, only enhanced.
    1. If done right this might even be backwards compatible, though this is clearly not in scope and I don't care about backwards compatibility at this point
1. Using this core-linking-functionality one can implement other `libimag*`s which simply use the core functionality from the linking library to implement "custom linking" in other places in the header, for example in `mail.links.*`.

**but**:

If we do it like the latter point from above, we lose the ability to trace links.

Simple use-case: One wants to find out how Entry A is linked to Entry B (find the link-path between them). With the current setup, we only would need to find a way through the net of linked entries.

If we implement custom linking in custom places inside the header, we would need to be able to find links in the header and do this based on them. This is simply not do-able, IMHO.

A fix for this problem that I could think of would be that we keep the linking as-is (plus annotations as mentioned above) and give the libraries the ability to annotate links with custom data in their part of the header. E.G.:

```toml
[imag]
links = [ "foo/bar"]

[mail]
[mail.link_annotations]
[[annotation]]
annotated = "foo/bar"
annotation = "thread_msgid"
```

Where `[[annotation]]` is a single annotation (from an array of annotations) and `annotation = "thread_msgid"` is simply a hint to the library that `"foo/bar"` (as in `annotated = "foo/bar"`) is a link to another mail from the thread.

We still might run into problems if one link vanishes and the annotation get invalidated by that. But that are edge-cases we have to deal with anyways, I guess.

I'm not sure I like that approach at all...

---

The second idea would be that modules can annotate links with custom data in the link itself (same approach like above with user-annotations), but in a sub-table they can manage on their own. E.G.:

```toml
[imag]
links = [
  { link = "foo/bar",
    annotations = {
      mail = { custom = "data", from = "the mail module", }
    }
  }
```

The link object in the header could also contain `user_annotation = "String"` - no problem there.

I like that approach more than the one above, though it would blow up the implementation of `libimagentrylink` and make it rather complex IMHO.

---

Sidenote: Why are links stored in `[imag]` by the way? They really should be in `[link]`, shouldn't they?

**Comment by mario-kr** [comment]
ok, here's my senf:
Imag as it is, is already very ambitous. The feature as you describe it will complicate an easy-to-understand feature, the linking. I don't think this is needed right now. The aim should rather be to look at how imag will work with additional backends (like mails...) and then decide if links need to be extended. I actually want to remind you that imag does not want to duplicate functionality, and, to take your example, filtering all e-mails for a certain author sounds like a job for the e-mail manager/program to me.
Implementing this feature might actually result in named relationships rather than links.

To the sidenote: links are an internal feature of imag, so putting them in `[imag]` doesn't sound that bad to me

**Comment by neithernut** [comment]
> Yes, but the mentioned usecase was to retrieve related links, and if you have direct links in the store, you can do this much faster as loading possibly thousands of mails from a Maildir, just to find out that there are no related mails (read: more mails in the same thread) (or only few).

When we have to read the mails, we might need to `stat()`/`open()` twice the number of files. Parsing a mail header should actually be a bit less as complex as parsing TOML, so the CPU-time used should be a bit less than twice the amount, while the most time will be spent waiting for the disk.

However, a cache is multiple times faster anyway, so we should rather implement proper caching instead of polluting the store with custom caches which are a PITA to maintain. We already discussed caches a few months back, I think, and just this evening I was yet again thinking about caching. I'll look for an old issues and reopen it, or open a new one.

----

Regarding your other implementation proposals:

As you already wrote yourself, the first proposal introduces data-redundancy with all the known drawbacks while the second approach pretty much boosts the level of complexity and uglyness.
Also:

> If we do it like the latter point from above, we lose the ability to trace links.

I do not quite follow your reasoning here. If you have a special, module-specific entry serving as a kind of link-hub, you can pretty easily represent that hub as something like a category to the user. Finding all members of that category is just a matter of listing all links from that category. Finding all entries linked to an entry can be done by following the link to a category and then proceeding getting all the category's members.

Linking an entry to all other entries of the same category directly, which you propose, involves creating and managing that list in `N` places and you end up with `N*N` links. Your proposal is not as reasonable as you might think it is, at least in my opinion.

Last but not least, I agree with @mario-kr. As I wrote myself, links ought to be simple. And extending links *just* for very quirky search performance optimizations yielding only a speed-up of 2, maximum, is a no-go. Especially if there are other means of speeding up searched. Caches are the way to go.

**Comment by matthiasbeyer** [comment]
>>If we do it like the latter point from above, we lose the ability to trace links.

> I do not quite follow your reasoning here. If you have a special, module-specific entry serving as a kind of link-hub, you can pretty easily represent that hub as something like a category to the user.

I mean that it has to be find-able by low-level libraries (speaking of `libimagentrylink`). If we put links in other sub-headers than `imag.link` (or whereever it is atm), would make it hard to build a net of entries.

Extending the link functionality should _only_ be done if we add meta-information to the link entries, not by implementing _another_ link entry in some other place in the header.

That was my whole point there.

**Comment by neithernut** [comment]
>> Parsing a mail header should actually be a bit less as complex as parsing TOML, so the CPU-time used should be a bit less than twice the amount, while the most time will be spent waiting for the disk.
>
> You forget that the TOML header is already parsed in the moment we know where the mail file actually lives.

If you are searching for mails in your store you usually don't have them loaded yet. Hence, you end up reading those TOML files, too.

> Extending the link functionality should only be done if we add meta-information to the link entries, not by implementing another link entry in some other place in the header.

In this case this would be another argument for *not* extending links.

**Comment by matthiasbeyer** [comment]
> If you are searching for mails in your store you usually don't have them loaded yet. Hence, you end up reading those TOML files, too.

That's not true. When you search for mails in the store, the appropriate store files are already loaded. The module knows that there are references to mails somewhere under `/ref/`, but it does not know which ones are mails. We then load the Refs and check whether they are references to mails, and if they are, we have them. At that point, the `Ref` objects (including the parsed header) live completely in memory already.

**Comment by neithernut** [comment]
I still don't follow your reasoning. You have to load the entries at some point. If you run a query using a one-shot command, this *definitely* can be considered part of the search, at least form the user's perspective. Even with long-lived processes, it is doubtful that *all* the entries are loaded at *any* given point in time.

**Comment by matthiasbeyer** [comment]
>> Extending the link functionality should only be done if we add meta-information to the link entries, not by implementing another link entry in some other place in the header.

> In this case this would be another argument for not extending links.

So if I understand you right, you wouldn't attach meta information to links at all?

We still need a way to create some relationships between entries, which can have meta-information attached to them. Connecting two entries is one point, making the connection meaningful is another and I consider it really important.

So we clearly need to lay out our options to implement such a thing.


> I don't think this is needed right now. The aim should rather be to look at how imag will work with additional backends (like mails...) and then decide if links need to be extended. I actually want to remind you that imag does not want to duplicate functionality, and, to take your example, filtering all e-mails for a certain author sounds like a job for the e-mail manager/program to me.

That's true on one hand, but telling imag about relationships is another.

> Implementing this feature might actually result in named relationships rather than links.

A link is a relationship, in some way. Making them annotateable is just an extension and 

>  The feature as you describe it will complicate an easy-to-understand feature, the linking.

The interface for linking shouldn't change with this new feature (the user interface/commandlineinterface won't change at all). It will "simply" be extended in the file format and in the library interfaces.

**Comment by neithernut** [comment]
> We still need a way to create some relationships between entries, which can have meta-information attached to them. Connecting two entries is one point, making the connection meaningful is another and I consider it really important.

I think we have a problem here with the core semantics of links being completely unclear. IIRC the entire concept of a link was introduced for referencing external content (a file) from an entry, or to extend that functionality. Linking to other entries was an *additional* feature. I don't even remember about the ideas behind it.

For links to external files/URIs, a link has a clear semantic. The resource referred to is a (complementary) source for information relevant to the entry. The details are obvious from the context. If, for example, an entry representing an appointment refers to a iCalendar file or a webdav, it clearly declares that resource a source from which the entry was created and/or with which it may be synced.

I assumed that the semantics of internal links could, at least, also be deferred from their context, the type of the entries involved that is. A link between an entry and a note, for example, has a pretty clear semantic imo. So has a link between a music album and a track. However, you now insist that there are use-cases where links *must* carry metadata to create the necessary semantics and provide examples which do not make any sense to me.

So what are links, semantically, to you? Maybe you think of links like associations in an object model? In this case, *maybe* we could extend links with a designator on each side. Modules could then specifically look for links with a certain semantic (e.g. a link designated as "album" in an entry representing a "track"). Also, you can navigate through those entries like one would do in OCL, fine-controlling which links to follow and which not. If you *really* need additional metadata for a group of entries, you use a module-specific intermediate entry.

**Comment by matthiasbeyer** [comment]
>  IIRC the entire concept of a link was introduced for referencing external content (a file) from an entry, or to extend that functionality. Linking to other entries was an additional feature. 

Nope. The idea behind the links was to be able to connect two imag-internal entries. Links to outside content was not intended. Links to UR_L_s was, though.

Linking store-external content was an additional feature which has absolutely nothing to do with the conecpt of linking  (and is hence called _references_/_ref_).

> However, you now insist that there are use-cases where links must carry metadata to create the necessary semantics and provide examples which do not make any sense to me.

If you link two mails, what can you infer from this type of linking? Are these two messages linked because they belong to one discussion (not necessarily one email thread or something like this)? Are they linked because the user linked them as she thinks these are related? Are they linked because they come from the same group of people (so not from the same contact/mail adress/etc, but somehow from a group where people are in some way related to another)? 

You simply don't know. This link could be automatically inferred by imag because (future talk here) imag inferred that these two mails are related in some way, maybe it was created by the user because of some thing we don't know about.

> So what are links, semantically, to you? 

Links are nothing more than a relation between pieces of data. The _type_ of the relation would be the additional attribute.

> Maybe you think of links like associations in an object model? 

You could put it this way, yes, although not with an ownership-kind-of-semantic.

> Also, you can navigate through those entries like one would do in OCL, _fine-controlling which links to follow and which not._

Yes, this fine-controlling when traversing is one key feature that comes with annotated links that I have in mind.

> If you really need additional metadata for a group of entries, you use a module-specific intermediate entry.

That's hte point I'm not sure about. Clearly the data has to be stored in a way that other modules don't run into problems when reading/writing the header data. But on the other hand, we cannot do what you suggest above:

> _fine-controlling which links to follow and which not._

if we store the meta information in arbitrary places in the module-specific header part. E.G.:

If we store meta-information in this way:

```toml
[imag]
links = [ "some/track", "some/other" ]
[music]
link_type = "album"
```

We cannot use the `music.link_type` data when traversing, as the traversing code would have to be able to find this data. This is not possible if we traverse with another tool than `imag-music`, for example, as the other tool does not _know_ `music.link_type` and its possible values. Another E.G.:

```bash
imag view --filter-script "entry.type == 'music' && entry.belongs_to_album 'some'"
```

The above wouldn't be do-able from `imag-view`, as it does not know about how to resolve the "belongs to album"-part.

If we build the header in a way where other modules can find such data:

```toml
[imag]
links = [ {
   link = "some/track",
   meta = { music = { type = "album" } } },
   {
   link = "some/other",
   meta = { music = { type = "album" } } },
]
```

(exact format is to be defined, the above might not be valid TOML but I guess you can see my point)

We end up with a way to query this meta information in a rather entry-type-agnostic-way. The above commandline call would be:

```bash
imag view --filter-script "entry.type == music && entry.links.map(meta.music.type).all_eq('album')"
```

I guess you get the basic idea.

The above header contains the `meta` field in each `links` entry, which is a map where each module can  (under `meta.<modulename>`) store arbitrary data. This way, the `libimaglink` can be used to query this data and we wouldn't need to include `libimagfoo` to be able to filter `imag.links.0.meta.<foo>`.

---

Please note that the above `imag-view` calls are made up for example purposes and I didn't think much about how the interface would be nice. They are simply made verbose to show my point.

**Comment by neithernut** [comment]
> If you link two mails, what can you infer from this type of linking? Are these two messages linked because they belong to one discussion (not necessarily one email thread or something like this)? Are they linked because the user linked them as she thinks these are related? Are they linked because they come from the same group of people (so not from the same contact/mail adress/etc, but somehow from a group where people are in some way related to another)?
>
> You simply don't know. This link could be automatically inferred by imag because (future talk here) imag inferred that these two mails are related in some way, maybe it was created by the user because of some thing we don't know about.

This problem doesn't occur if you represent the "group" of messages using an entry for linking those together. It makes more sense anyway, because you'll virtually always want to add additional data to that group, e.g. notes to a discussion. Having all that information stuffed into links is plain ugly.

> That's hte point I'm not sure about. Clearly the data has to be stored in a way that other modules don't run into problems when reading/writing the header data. But on the other hand, we cannot do what you suggest above:
>
> [...]

Well, I thought of something like
```
[imag]
links = [ {ref = "some/track", type = "music::track" }, "some/other" ]
```

Links have exactly one type. Anything else doesn't make much sense imo. If multiple modules create links between two entries, they (logically) are not equal. Also, a link consisting of only a string would denote a link with that object, but with no type attached. This way, you are backwards compatible.

>     imag view --filter-script "entry.type == music && entry.links.map(meta.music.type).all_eq('album')"
>
> I guess you get the basic idea.

No, I absolutely don't. This query doesn't make any sense at all. No user (or a very small fraction) would do that. It's much more natural to look for an album/all songs in an album. The rather complex way you formulated that filter is also rather cumbersome. For example, why not add some syntactic sugar to the filter and let `<x>.<y>` mean to "all referenced objects referenced by links of type `<y>` in `<x>`" if `<y>` cannot be resolved otherwise? It's much more natural. This way, you could check whether a "track" belongs to an album with a plain `entry.music::album == <some album>` or `entry.music::album.contains(<some album>)`.

**Comment by matthiasbeyer** [comment]
> This problem doesn't occur if you represent the "group" of messages using an entry for linking those together.

That would be another option I didn't think of yet, yes. It would still lead to a _massive_ amount of entries in the store, though.

> Links have exactly one type. Anything else doesn't make much sense imo. If multiple modules create links between two entries, they (logically) are not equal. Also, a link consisting of only a string would denote a link with that object, but with no type attached. This way, you are backwards compatible.

That would be another possibility that sounds rather good to me, yes.

---

So lets adhere:

* A Link can have a `type`. Equality of links are defined as `(a.target + a.type) == (b.target + b.type)`.
* Grouping of entries (for example "all mails in thread `X`) are done via another entry which links to the entries in the group and adds more meta-information

---

One small idea on the second point: If we now traverse the net of entries, we end up having a lot of "intermediate" entries which are only there for linking and providing meta-information.

That could be a small inconvenience, though I guess we can work around that later, right?

**Comment by neithernut** [comment]
> That would be another option I didn't think of yet, yes. It would still lead to a massive amount of entries in the store, though.

Compared with the amount of entries being linked together, the number should be relatively small (at least for group sizes `> 2`). Also, the "group" entries only emerge if there is actually a need for them. And it's still better than duplicating meta-information all over the place.

> So lets adhere:
>
> [...]

agreed.

> One small idea on the second point: If we now traverse the net of entries, we end up having a lot of "intermediate" entries which are only there for linking and providing meta-information.
> 
> That could be a small inconvenience, though I guess we can work around that later, right?

I think in practice, the intermediate entries will end up not being a drawback when traversing the graph but a benefit. But if you want to have a graph where intermediate entries don't show up, you could easily mark them in some way, e.g. using a header-entry of some sort. This is, however, another issue.

**Comment by matthiasbeyer** [comment]
> But if you want to have a graph where intermediate entries don't show up, you could easily mark them in some way, e.g. using a header-entry of some sort. This is, however, another issue.

Yep, that's what I thought of and yes, that's another issue.

**Comment by matthiasbeyer** [comment]
Because we are now [moving off of github](https://github.com/matthiasbeyer/imag/pull/1505) I'm closing this PR/Issue now. The reasons why we do this was described in a [blogpost](https://imag-pim.org/blog/2018/06/10/off-of-github/). Feel free to resubmit this PR/Issue via the mailinglist.
